syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.six.bic.mc.appl.grpc";
option java_outer_classname = "MessageConverterServiceProto";

package com.six.bic.mc.appl.grpc;

service MessageConverterService {
    rpc Recognize(RecognizeMessageRequest) returns(RecognizeMessageResponse) {}
    rpc Convert(ConverterRequest) returns(ConverterResponse) {}
    rpc Validate(ValidateRequest) returns(ValidateResponse) {}
}

message ValidateRequest {
    string message = 1;
}

message ValidateResponse {
    bool isSuccess = 1;
    repeated string errors = 5;
}

message RecognizeMessageRequest {
    string message = 1;
    repeated string hints = 5;
}

message RecognizeMessageResponse {
    bool isSuccess = 1;
    repeated string errors = 5;
    string uri_msg = 10;
}

message ConverterRequest {
    string uri_msg = 1;
    string msg = 5;
    bool skip_validation = 10;
}

message ConverterResponse {
    bool isSuccess = 1;
    repeated ConversionError errors = 2;

    string uri_msg = 5;

    string msg = 10;
}

message ConversionError {
    enum Type {
        INTERNAL_ERROR = 0;
        PARSER_ERROR = 1;
        MISSING_FIELD_ERROR = 2;
        FIELD_FORMAT_ERROR = 3;
        VALIDATION_ERROR = 4;
        UNSUPPORTED = 5;
    }

    string message = 1;
    Type type = 2;
}